connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java

```diff
@@ -109,6 +121,69 @@ public void testThreadName() {
+        serializedPartitionOffsets.put(
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Arrays.asList("connector1", Collections.singletonMap("partitionKey", "partitionValue1")))),
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Collections.singletonMap("offsetKey", "offsetValue")))
+        );
+        serializedPartitionOffsets.put(
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Arrays.asList("connector1", Collections.singletonMap("partitionKey", "partitionValue1")))),
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Collections.singletonMap("offsetKey", "offsetValue2")))
+        );
+        serializedPartitionOffsets.put(
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Arrays.asList("connector1", Collections.singletonMap("partitionKey", "partitionValue2")))),
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Collections.singletonMap("offsetKey", "offsetValue")))
+        );
+        serializedPartitionOffsets.put(
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Arrays.asList("connector2", Collections.singletonMap("partitionKey", "partitionValue")))),
+                ByteBuffer.wrap(jsonConverter.fromConnectData("", null,
+                        Collections.singletonMap("offsetKey", "offsetValue")))
+        );
+
+        store.set(serializedPartitionOffsets, setCallback).get();
+        store.stop();
+
+        // Restore into a new store to ensure correct reload from scratch
+        FileOffsetBackingStore restore = new FileOffsetBackingStore(jsonConverter);
+        restore.configure(config);
+        restore.start();
+
+        Set<Map<String, Object>> connectorPartitions1 = restore.connectorPartitions("connector1");
+        assertEquals(2, connectorPartitions1.size());
+
+        Set<Map<String, Object>> expectedConnectorPartition1 = new HashSet<>();
+        expectedConnectorPartition1.add(Collections.singletonMap("partitionKey", "partitionValue1"));
+        expectedConnectorPartition1.add(Collections.singletonMap("partitionKey", "partitionValue2"));
+
+        assertTrue(connectorPartitions1.containsAll(expectedConnectorPartition1));
+        assertTrue(expectedConnectorPartition1.containsAll(connectorPartitions1));
+
+        Set<Map<String, Object>> connectorPartitions2 = restore.connectorPartitions("connector2");
+        assertEquals(1, connectorPartitions2.size());
+
+        Set<Map<String, Object>> expectedConnectorPartition2 = Collections.singleton(Collections.singletonMap("partitionKey", "partitionValue"));
+
+        assertTrue(connectorPartitions2.containsAll(expectedConnectorPartition2));
+        assertTrue(expectedConnectorPartition2.containsAll(connectorPartitions2));
```

Comment:
Worth adding a case after here where we emit a null value for one of the source partitions already present in the store and verify that it gets wiped?

Reply:
Makes sense, added. This also made me realize that there was a bug in the `FileOffsetBackingStore` implementation for `connectorPartitions`. We were only populating the `connectorPartitions` map in the `load` method which only happens once during startup; we should also be populating it on calls to `save` which gets called on every `OffsetBackingStore::set` (this wasn't an issue in the `KafkaOffsetBackingStore` where the `connectorPartitions` map is populated in the separate thread consuming from the `KafkaBasedLog`).